manabi/filesystem.py

```diff
@@ -142,24 +150,167 @@ def __init__(
+
+        We can't call `super()` here, because we need to use `open` from `smart_open`.
+        """
+        assert not self.is_collection
+        return s3_open(
+            f"s3://{self.bucket_name}/{self.file_path}",
+            "rb",
+            transport_params={"client": self.s3},
+        )
+
+    def begin_write(self, *, content_type):
+        """
+        Open content as a stream for writing.
+
+        We can't call `super()` here, because we need to use `open` from `smart_open`.
+        """
+        self.process_pre_write_hooks()
+        assert not self.is_collection
+        if self.provider.readonly:
+            raise DAVError(HTTP_FORBIDDEN)
+        return s3_open(
+            f"s3://{self.bucket_name}/{self.file_path}",
+            "wb",
+            transport_params={"client": self.s3},
+        )
+
+
+class ManabiS3Provider(ManabiProvider):
+    def __init__(
+        self,
+        root_folder,
+        endpoint_url,
+        aws_access_key_id,
+        aws_secret_access_key,
+        region_name,
+        bucket_name,
+        readonly=False,
+        shadow=None,
+        cb_hook_config: Optional[CallbackHookConfig] = None,
+    ):
+        super(FilesystemProvider, self).__init__()
+
+        if not root_folder:
+            raise ValueError("Invalid root path: {dir}".format(dir=root_folder))
+
+        self.root_folder_path = str(root_folder)
+        self.readonly = readonly
+        if shadow:
+            self.shadow = {k.lower(): v for k, v in shadow.items()}
+        else:
+            self.shadow = {}
+
+        self.fs_opts = {}
+        # Get shadow map and convert keys to lower case
+        self.shadow_map = self.fs_opts.get("shadow_map") or {}
+        if self.shadow_map:
+            self.shadow_map = {k.lower(): v for k, v in self.shadow_map.items()}
+
+        self._cb_hook_config = cb_hook_config
+
+        self.endpoint_url = endpoint_url
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.region_name = region_name
+        self.bucket_name = bucket_name
+        self.s3 = get_boto_client(
+            endpoint_url=self.endpoint_url,
+            aws_access_key_id=self.aws_access_key_id,
+            aws_secret_access_key=self.aws_secret_access_key,
+            region_name=self.region_name,
+        )
+        self._file_resource = None
+
+    def get_file_resource(self, path, environ, fp):
+        try:
+            return ManabiS3FileResource(
+                self.s3,
+                self.bucket_name,
+                path,
+                environ,
+                fp,
+                cb_hook_config=self._cb_hook_config,
+            )
+        except ClientError as ex:
+            if ex.response["Error"]["Code"] == "NoSuchKey":
+                # File does not exist
+                return None
+
+    def file_exists(self, path: Path):
+        # Is handled implicitly in `get_file_resource()`
+        return True
```

Comment:
My need to conform the protocol (even though not strictly needed), would probably make do something like:

```python
return bool(self.s3.get_object(Bucket=self.bucket_name, Key=path))
```

any downsides to this? (I haven't checked the interface get_object, maybe it throws exceptions if the file does not exists). Is it a performance thing? In that case we might cache s3-object?

Reply:
I've refactored the `file_exists()` logic into the `get_file_resource()` method.

Reply:
I read ` def do_PROPFIND` in wsgi-dav and I see now that this works properly. But I did not find `file_exists()` in wsgi-dav at all. I think it got removed at some point. So you should be able to remove `def file_exists()` completely.

Reply:
`file_exists()` was an approach by me, to handle different types of paths (fs vs s3). But it's already removed, as per my commend above:

> I've refactored the file_exists() logic into the get_file_resource() method.